openshift-kube-apiserver/enablement/intialization.go

```diff
@@ -83,3 +98,69 @@ func ForceGlobalInitializationForOpenShift() {
 	// SkipSystemMastersAuthorizer disable implicitly added system/master authz, and turn it into another authz mode "SystemMasters", to be added via authorization-mode
 	server.SkipSystemMastersAuthorizer()
 }
+
+var SCCAdmissionPlugin = sccadmission.NewConstraint()
+
+// we patch this method so that we can mutate the returned podspec to indicate the pod that will exist AFTER SCC mutation
+// we do this by taking the external pod, converting to internal, building a synthetic podsecurityadmission attributes, passing to SCC,
+// checking for error and returning the mutated podspec.
+func extractPodSpecFromTemplateWithSCCMutation(obj runtime.Object, template *corev1.PodTemplateSpec) (*metav1.ObjectMeta, *corev1.PodSpec, error) {
```

Comment:
I believe a better solution would be to create our own `SCCMutatingExtractor` such as
```go
type SCCMutatingExtractor struct {
    sccAdmission admission.MutationInterface
    kubePodExtractor podsecurityadmission.PodExtractor
}
```

where in a `NewSCCMutatingExtractor()` you supply the sccadmission mutator and the `podsecurityadmission.DefaultPodExtractor`.

You then implement the `SCCMutatingExtractor.ExtractPodSpec()` method as this `extractPodSpecFromTemplateWithSCCMutation()` but instead of patching the upstream extractor to be able to use its extraction method `podsecurityadmission.ExtractPodSpecFromTemplate_original()`, you just directly call `sccMutatingExtractor.kubePodExtractor.ExtractPodSpec()`.

All the above could live in apiserver-library-go, reducing the changes to minimum.

You then only need to supply the PodExtractor of your own in https://github.com/openshift/kubernetes/blob/ed9338086f74c836858cb5ed50dab6bf1fcc6b9e/plugin/pkg/admission/security/podsecurity/admission.go#L119

This way you keep the carry patch to a minimum, which should also make @soltysh happy.

I hope I did not forget any OpenShift patching particularity which would make this not possible.

Reply:
Updated